src/iminuit/qtwidget.py

```diff
@@ -0,0 +1,413 @@
+                f"FCN = {fval:.3f}",
+                transform=self.fig.get_axes()[0].transAxes,
+                fontsize="x-large",
+            )
+            if from_fit and report_success:
+                self.fig.get_axes()[-1].text(
+                    0.95,
+                    1.05,
+                    f"{'success' if minuit.valid and minuit.accurate else 'FAILURE'}",
+                    transform=self.fig.get_axes()[-1].transAxes,
+                    fontsize="x-large",
+                    ha="right",
+                )
+
+        def fit(self):
+            if self.algo_choice.currentText() == "Migrad":
+                minuit.migrad()
+            elif self.algo_choice.currentText() == "Scipy":
+                minuit.scipy()
+            elif self.algo_choice.currentText() == "Simplex":
+                minuit.simplex()
+                return False
+            else:
+                assert False  # pragma: no cover, should never happen
+            return True
+
+        def on_parameter_change(self, from_fit=False, report_success=False):
+            if any(x.fit.isChecked() for x in self.parameters):
+                saved = minuit.fixed[:]
+                for i, x in enumerate(self.parameters):
+                    minuit.fixed[i] = not x.fit.isChecked()
+                from_fit = True
+                report_success = self.do_fit(plot=False)
+                self.results_text.clear()
+                self.results_text.setHtml(minuit._repr_html_())
+                minuit.fixed = saved
+            elif from_fit:
+                self.results_text.clear()
+                self.results_text.setHtml(minuit._repr_html_())
+            else:
+                self.results_text.clear()
+
+            for ax in self.fig.get_axes():
+                ax.clear()
+            self.plot_with_frame(from_fit, report_success)
+            self.canvas.draw_idle()
+
+        def do_fit(self, plot=True):
+            report_success = self.fit()
+            for i, x in enumerate(self.parameters):
+                x.reset(val=minuit.values[i])
+            if not plot:
+                return report_success
+            self.on_parameter_change(from_fit=True, report_success=report_success)
+
+        def on_update_button_clicked(self):
+            for x in self.parameters:
+                x.slider.setTracking(self.update_button.isChecked())
+
+        def on_reset_button_clicked(self):
+            minuit.reset()
+            minuit.values = original_values
+            minuit.limits = original_limits
+            for i, x in enumerate(self.parameters):
+                x.reset(val=minuit.values[i], limits=original_limits[i])
+            self.on_parameter_change()
+
+    if qt_exec:
+        app = QtWidgets.QApplication.instance()
+        if app is None:
+            app = QtWidgets.QApplication([])
+        main_window = MainWindow()
+        main_window.show()
+        app.exec()
+    else:
+        return MainWindow()
```

Comment:
I haven't tested that, but it looks like this code can be made to work without qt_exec.
```suggestion
    app = QtWidgets.QApplication.instance()
    run_event_loop = False
    if app is None:
         app = QtWidgets.QApplication([])
         run_event_loop = True

    main_window = MainWindow()
    main_window.show()

    if run_event_loop:
        app.exec() # this blocks the main thread
    return main_window
```

That should run the event loop if it does not exist when we call make_widget, but not if the event loop is already running (such as in the tests).

Reply:
If this does not work, then let's keep the original solution, but rename `qt_exec` to something more sensible, like `start_event_loop = True`.

In the end, this parameter solely exists for the unit tests, right? That was not immediately clear when I read the code.

Reply:
https://github.com/scikit-hep/iminuit/blob/2813c5131db0e6fe1024ba39d3725937fdc9bb31/src/iminuit/qtwidget.py#L381-L393
My solution would now be this. My idea was, that if no event loop is running, the user probably wants the app to be started by the interactive call and block further code execution in order to debug some fit. To enable multiple interactive calls in one script, I set a name and check if the existing instance has that name. 
I additionally wanted a way for the interactive call to return the widget without starting an event loop for two reasons: One is the tests and the more important second reason is, that one could easily include the widget in a another application.
Also, by using the qtbot fixture in the tests, the name of this qtbot QApplication instance should be different and we should get the widget returned.

Reply:
Ok, I see. Using the application name to change behavior here is very brittle, let's not do that. In that case, I rather accept your original code with the keyword qt_exec, but please replace it with "run_event_loop=True", so the default is to run the event loop.

Reply:
Ok, I added the run_event_loop argument, but did not add it to the docstring.

Reply:
Ok, wait a minute, why do we have to expose "run_event_loop" to the user at all in Minuit.interactive? We don't need that for testing, I believe? I would like to avoid exposing run_event_loop to the top level user interface. Not documenting the keyword is not a good option, undocumented keywords exposed by an API are bad.

Can you make your use-case work of including the widget in another application without adding that to Minuit.interactive? Can't you call make_widget directly in your other application? It is ok that make_widget has that option, because that is a semi-private API that the user is not expected to know.

Reply:
Ah okay, I understand now. Yes, that will work for me. I changed the tests in test_draw.py so that they use make_widget directly.

Reply:
Cool, thanks.